MaxText/layers/gpt3.py

```diff
@@ -0,0 +1,340 @@
+"""
+ Copyright 2023 Google LLC
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
+"""Transformer model definition."""
+# pylint: disable=arguments-differ
+# pylint: disable=no-name-in-module
+
+
+from typing import Any, Tuple
+
+from jax.sharding import Mesh
+from jax import lax
+import jax.numpy as jnp
+from jax.ad_checkpoint import checkpoint_name
+
+from flax import linen as nn
+
+from layers import attentions
+from layers import initializers
+from layers import linears
+from layers import models
+
+AttentionOp = attentions.AttentionOp
+
+import common_types
+
+Array = common_types.Array
+Config = common_types.Config
+DType = common_types.DType
+Mesh = common_types.Mesh
+AxisNames = common_types.AxisNames
+BATCH = common_types.BATCH
+LENGTH = common_types.LENGTH
+HEAD = common_types.HEAD
+D_KV = common_types.D_KV
+
+DenseGeneral = linears.DenseGeneral
+NdInitializer = initializers.NdInitializer
+Initializer = initializers.Initializer
+nd_dense_init = initializers.nd_dense_init
+
+
+#-----------------------------------------
+# The Normalization Layer specific for GPT3
+#-----------------------------------------
+
+class Gpt3LayerNorm(nn.Module):
```

Comment:
Why is this layer norm different from the existing Layer norm?

Reply:
In pax `Gpt3LayerNorm`: (inputs - mean) * rsqrt(mean(square(inputs - mean)) + self.epsilon) * (scale + 1) + bias
In maxtext the exisitng `rms_norm` : inputs * rsqrt(mean(square(inputs)) + self.epsilon) * scale
see [doc](https://docs.google.com/spreadsheets/d/1lUAayl6dcL1dMKcCp2Uzo730YK_k-AeCIUq8jfrGZ84/edit#gid=0)

Other than that, rms_norm is calculating in fp32 in reduction. Have to add lots of new flags without a customized gpt3 implementation.

Reply:
Ah I forget we don't have a LayerNorm.